velox/common/memory/MmapAllocator.h

```diff
@@ -0,0 +1,226 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <array>
+#include <atomic>
+#include <cstdint>
+#include <memory>
+#include <mutex>
+#include <unordered_set>
+
+#include "velox/common/base/BitUtil.h"
+#include "velox/common/memory/MappedMemory.h"
+
+namespace facebook::velox::memory {
+
+// Denotes a number of pages of one size class, i.e. one page consists
+// of a size class dependent number of consecutive machine pages.
+using ClassPageCount = int32_t;
+
+struct MmapAllocatorOptions {
+  //  Capacity in bytes, defult 512MB
+  uint64_t capacity = 1L << 29;
+};
+// Implemantation of MappedMemory with mmap and madvise. Each size
+// class is mmapped for the whole capacity. Each size class has a
+// bitmap of allocated entries and entries that are backed by
+// memory. If a size class does not have an entry that is free and
+// backed by memory, we allocate an entry that is free and we advise
+// away pages from other size classes where the corresponding entry
+// is free. In this way, any combination of sizes that adds up to
+// the capacity can be allocated without fragmentation. If a size
+// needs to be allocated that does not correspond to size classes,
+// we advise away enough pages from other size classes to cover for
+// it and then make a new mmap of the requested size
+// (ContiguousAllocation).
+class MmapAllocator : public MappedMemory {
+ public:
+  explicit MmapAllocator(const MmapAllocatorOptions& options);
+
+  virtual ~MmapAllocator() = default;
+
+  bool allocate(
+      MachinePageCount numPages,
+      int32_t owner,
+      Allocation& out,
+      std::function<void(int64_t)> beforeAllocCB = nullptr,
+      MachinePageCount minSizeClass = 0) override;
+
+  int64_t free(Allocation& allocation) override;
+
+  bool allocateContiguous(
+      MachinePageCount numPages,
+      Allocation* collateral,
+      ContiguousAllocation& allocation,
+      std::function<void(int64_t)> beforeAllocCB = nullptr) override;
+
+  void freeContiguous(ContiguousAllocation& allocation) override;
+
+  // Checks internal consistency of allocation data
+  // structures. Returns true if OK.
+  bool checkConsistency() override;
+
+  MachinePageCount capacity() const {
+    return capacity_;
+  }
+
+  MachinePageCount numAllocated() const override {
+    return numAllocated_;
+  }
+
+  MachinePageCount numMapped() const override {
+    return numMapped_;
+  }
+
+  std::string toString() const override;
+
+ private:
+  static constexpr uint64_t kAllSet = 0xffffffffffffffff;
+
+  class SizeClass {
+   public:
+    SizeClass(size_t capacity, MachinePageCount pageSize);
```

Comment:
MachinePageCount pageSize is confusing. MachinePageCount is a type for number of pages, but pageSize sounds like it is the size of the page. Would you clarify?